Update challenges to match the spec

bifurcation opened the pull request:
Fixes #434

kuba commented:
I'm getting `http: panic serving 127.0.0.1:34805: Tried to serialize a nil pointer.` and socket gets closed.

Relevant client log (local dev branch):

```
2015-07-19 13:17:20,156:DEBUG:acme.client:Serialized JSON: {"type": "dvsni", "validation": {"header": {"alg": "RS256", "jwk": {"e": "AQAB", "kty": "RSA", "n": "31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL0t
xiX9zFDM_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstqK41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBVMX
Ff4w050S99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQQsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ"}}, "payload": "eyJ0b2tlbiI6ICJVZVVFMktwNTFjbDMwSi1UcGprM0hHQ2ttaXB3YlItdmdNU1I5WX
F4bHBzIiwgInR5cGUiOiAiZHZzbmkifQ", "signature": "Flb7l_75d_cOw6b-jHRyRkAxaWup8VpWcg2x9FFuImYXI86guRYWBBBHscSv9K31bsp4Y_mz8M8WIO5otc5lS-cqH9Ryw-d0tsmE8Zlb6OSjzbZWiYovBFmd_6wiTP3sLhAYTCXKhXkZgyFvIl
u2tQkyqtqHO3K1MEyu42YVsHUs47sbQYbc0LuLn6oi3b_CdlVPD4T01FP6BFIoSYaBBKWVZrfdY4YP5yE7z0KLvGfVLTBvvmg4-HDygBiXgJOY3jK05TCaWL5GfIAptwHatEYGQaaPpSP_I4t82DZURxlarR5EX-u_TOl9K0hkMHKvwyg0GlVYNNDwIgBvesWzI
Q"}, "resource": "challenge"}
2015-07-19 13:17:20,157:DEBUG:acme.jose.json_util:Omitted empty fields: x5c=(), alg=None, typ=None, crit=(), jku=None, cty=None, x5t=None, kid=None, x5tS256=None, x5u=None, jwk=None
2015-07-19 13:17:20,158:DEBUG:acme.jose.json_util:Omitted empty fields: x5c=(), crit=(), typ=None, jku=None, nonce=None, cty=None, x5t=None, kid=None, x5tS256=None, x5u=None
2015-07-19 13:17:20,158:DEBUG:root:Sending POST request to http://localhost:4000/acme/authz/KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78?challenge=1
```


Server log:
```
Server running, listening on 127.0.0.1:4000...
2015/07/19 13:17:19 Terminated request JSON={"ID":"rz2uGbLdnlFQsYmFKATPsAY5hm6oDe53t9SPTNi0KYE","Endpoint":"/acme/new-reg","Method":"HEAD","RequestTime":"2015-07-19T13:17:19.744466291Z","Response
Time":"2015-07-19T13:17:19.744552089Z","Error":"Method not allowed"}
2015/07/19 13:17:19 [SQL:   SELECT * FROM registrations WHERE jwk = ? 1:"{\"kty\":\"RSA\",\"n\":\"31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL0txiX9zFDM_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstq
K41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBVMXFf4w050S99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQ
QsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ\",\"e\":\"AAEAAQ\"}"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   SELECT * FROM registrations WHERE jwk = ? 1:"{\"kty\":\"RSA\",\"n\":\"31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL0txiX9zFDM_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstq
K41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBVMXFf4w050S99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQ
QsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ\",\"e\":\"AAEAAQ\"}"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   begin; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   insert into "registrations" ("id","jwk","contact","agreement","LockCol") values (null,?,?,?,?); 1:"{\"kty\":\"RSA\",\"n\":\"31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL0txiX9zFDM
_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstqK41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBVMXFf4w050S
99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQQsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ\",\"e\":\"AAEAAQ\"}" 2:"null" 3:"" 4:1]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   commit; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 Successful request JSON={"ID":"3LzRd6zgRr2wbqGxkECuZFR33AK8VHzZwC9MfvVdfro","Endpoint":"/acme/new-reg","Method":"POST","RequestTime":"2015-07-19T13:17:19.788815838Z","Response
Time":"2015-07-19T13:17:19.790043013Z","Requester":1}
2015/07/19 13:17:19 [SQL:   SELECT * FROM registrations WHERE jwk = ? 1:"{\"kty\":\"RSA\",\"n\":\"31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL0txiX9zFDM_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstq
K41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBVMXFf4w050S99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQ
QsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ\",\"e\":\"AAEAAQ\"}"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   begin; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   SELECT count(*) FROM registrations WHERE id = ? 1:1]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   update "registrations" set "jwk"=?, "contact"=?, "agreement"=?, "LockCol"=? where "id"=? and "LockCol"=?; 1:"{\"kty\":\"RSA\",\"n\":\"31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL
0txiX9zFDM_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstqK41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBV
MXFf4w050S99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQQsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ\",\"e\":\"AAEAAQ\"}" 2:"null" 3:"http://localhost:4300/terms" 4:2 5:1 6:1]%!s(MI
SSING) [%!s(MISSING)]
2015/07/19 13:17:19 [SQL:   commit; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:19 Successful request JSON={"ID":"kh7IORA7XxjeukPsbTXe9stWIWvOAHoHqhYJOuzgVTM","Endpoint":"/acme/reg/1","Method":"POST","RequestTime":"2015-07-19T13:17:19.796685265Z","ResponseTi
me":"2015-07-19T13:17:19.797783264Z","Requester":1}
2015/07/19 13:17:19 [SQL:   SELECT * FROM registrations WHERE jwk = ? 1:"{\"kty\":\"RSA\",\"n\":\"31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL0txiX9zFDM_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstq
K41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBVMXFf4w050S99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQ
QsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ\",\"e\":\"AAEAAQ\"}"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [AUDIT] Checked CAA records for le1.wtf, registration ID 1 [Present: false, Valid for issuance: true]
2015/07/19 13:17:20 [SQL:   begin; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   SELECT count(*) FROM pending_authz WHERE id = ? 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   SELECT count(*) FROM authz WHERE id = ? 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   insert into "pending_authz" ("id","identifier","registrationID","status","expires","challenges","combinations","LockCol") values (?,?,?,?,?,?,?,?); 1:"KoPKh14GMYNlkPwi
AeFp1uTgxy0JUbindXAJrZh6E78" 2:"{\"type\":\"dns\",\"value\":\"le1.wtf\"}" 3:1 4:"pending" 5:<nil> 6:"null" 7:"[[0],[1],[2]]" 8:1]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   commit; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   begin; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   SELECT count(*) FROM authz WHERE id = ? 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   SELECT count(*) FROM pending_authz WHERE id = ? 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   select "id","identifier","registrationID","status","expires","challenges","combinations","LockCol" from "pending_authz" where "id"=?; 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUb
indXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   update "pending_authz" set "id"=?, "identifier"=?, "registrationID"=?, "status"=?, "expires"=?, "challenges"=?, "combinations"=?, "LockCol"=? where "id"=? and "LockCol
"=?; 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78" 2:"{\"type\":\"dns\",\"value\":\"le1.wtf\"}" 3:1 4:"pending" 5:<nil> 6:"[{\"type\":\"simpleHttp\",\"status\":\"pending\",\"uri\":\"http://loca
lhost:4000/acme/authz/KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78?challenge=0\",\"token\":\"R412OrMiYFXOq7KsLGkbRtgWAHqbca5XC1ErBupM-uE\",\"tls\":true},{\"type\":\"dvsni\",\"status\":\"pending\",
\"uri\":\"http://localhost:4000/acme/authz/KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78?challenge=1\",\"token\":\"UeUE2Kp51cl30J-Tpjk3HGCkmipwbR-vgMSR9Yqxlps\"},{\"type\":\"dns\",\"status\":\"pend
ing\",\"uri\":\"http://localhost:4000/acme/authz/KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78?challenge=2\",\"token\":\"tkHXRGtj6gtSHvDZiqn1ILgg6D_3BLspx1soRkBLaEk\"}]" 7:"[[0],[1],[2]]" 8:2 9:"Ko
PKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78" 10:1]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   commit; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 Successful request JSON={"ID":"9e0S9_P55FMSp0RGHZbRXArPEQXuz65iX3XXxyEnNTI","Endpoint":"/acme/new-authz","Method":"POST","RequestTime":"2015-07-19T13:17:19.88751592Z","Respons
eTime":"2015-07-19T13:17:20.129802401Z","Requester":1,"Extra":{"AuthzID":"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78","Identifier":{"type":"dns","value":"le1.wtf"}}}
2015/07/19 13:17:20 [SQL:   begin; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   select "id","identifier","registrationID","status","expires","challenges","combinations","LockCol" from "pending_authz" where "id"=?; 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUb
indXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   commit; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   SELECT * FROM registrations WHERE jwk = ? 1:"{\"kty\":\"RSA\",\"n\":\"31iyzQz2hsLjkAN1JPd1lUHIYFeaRy9UL0txiX9zFDM_VHYzTYCEJaiV0X7o0d56rykFRfoeXdwa5FDMxCjmcXveRAIqtxstq
K41BLyf-nuA6-dTbIwN5imnUS1cgUdkS5SddLqwtRQQcuQxYYfwy4dOquCcEr4GJMAImrK_LW5PwfPyCL9E9c4lSKeKGeMkV7qMvgpj_HN65gO9vK0k2nSGV1IMaU8TxoBVMXFf4w050S99Zea3_iR8WFJ9tV7RgGRTMlL2MAGeji1N99433S_MQVU3lxtx4_PQ
QsxAEffAsHPKWLSvxZI7Znj1UHDLbN1R5GR2T-b3V_27pjBFMQ\",\"e\":\"AAEAAQ\"}"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   begin; ]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   SELECT count(*) FROM authz WHERE id = ? 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   SELECT count(*) FROM pending_authz WHERE id = ? 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 [SQL:   select "id","identifier","registrationID","status","expires","challenges","combinations","LockCol" from "pending_authz" where "id"=?; 1:"KoPKh14GMYNlkPwiAeFp1uTgxy0JUb
indXAJrZh6E78"]%!s(MISSING) [%!s(MISSING)]
2015/07/19 13:17:20 Successful request JSON={"ID":"-GWeTFXwmNjbLeLbvW0Zub-6Updz8mmS3C4gSCM5HlE","Endpoint":"/acme/authz/KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78?challenge=1","Method":"POST","R
equestTime":"2015-07-19T13:17:20.160744079Z","ResponseTime":"2015-07-19T13:17:20.163072223Z","Extra":{"AuthorizationExpires":null,"AuthorizationID":"KoPKh14GMYNlkPwiAeFp1uTgxy0JUbindXAJrZh6E78","
AuthorizationIdentifier":{"type":"dns","value":"le1.wtf"},"AuthorizationRegistrationID":1,"AuthorizationStatus":"pending"}}
2015/07/19 13:17:20 http: panic serving 127.0.0.1:34805: Tried to serialize a nil pointer.
goroutine 15 [running]:
net/http.func·011()
        /usr/lib/go/src/net/http/server.go:1130 +0xbb
encoding/json.func·002()
        /usr/lib/go/src/encoding/json/encode.go:265 +0x101
github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/square/go-jose.mustSerializeJSON(0x978840, 0x0, 0x0, 0x0, 0x0)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/square/go-jose/encoding.go:64 +0x168
github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/square/go-jose.JsonWebSignature.FullSerialize(0xc208441720, 0x49, 0x4b, 0xc2082a8240, 0x1, 0x1, 0x0, 0x0)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/square/go-jose/jws.go:219 +0xf7
github.com/letsencrypt/boulder/core.AcmeJWS.MarshalJSON(0xc208441720, 0x49, 0x4b, 0xc2082a8240, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/core/objects.go:397 +0x6f
github.com/letsencrypt/boulder/core.(*AcmeJWS).MarshalJSON(0xc20829d980, 0x0, 0x0, 0x0, 0x0, 0x0)
        <autogenerated>:5 +0xd3
encoding/json.marshalerEncoder(0xc20806f080, 0x99b980, 0xc2082a0d68, 0xd6, 0x0)
        /usr/lib/go/src/encoding/json/encode.go:416 +0x146
encoding/json.(*structEncoder).encode(0xc208469560, 0xc20806f080, 0x9d8460, 0xc2082a0cb8, 0xd9, 0x0)
        /usr/lib/go/src/encoding/json/encode.go:584 +0x2ba
encoding/json.*structEncoder.(encoding/json.encode)·fm(0xc20806f080, 0x9d8460, 0xc2082a0cb8, 0xd9, 0x9d8400)
        /usr/lib/go/src/encoding/json/encode.go:598 +0x58
encoding/json.(*arrayEncoder).encode(0xc208038558, 0xc20806f080, 0x8903c0, 0xc2080d16e0, 0x57, 0x0)
        /usr/lib/go/src/encoding/json/encode.go:687 +0x11e
encoding/json.*arrayEncoder.(encoding/json.encode)·fm(0xc20806f080, 0x8903c0, 0xc2080d16e0, 0x57, 0x890300)
        /usr/lib/go/src/encoding/json/encode.go:694 +0x58
encoding/json.(*sliceEncoder).encode(0xc208038560, 0xc20806f080, 0x8903c0, 0xc2080d16e0, 0x57, 0x0)
        /usr/lib/go/src/encoding/json/encode.go:664 +0xb4
encoding/json.*sliceEncoder.(encoding/json.encode)·fm(0xc20806f080, 0x8903c0, 0xc2080d16e0, 0x57, 0xc2080d1600)
        /usr/lib/go/src/encoding/json/encode.go:673 +0x58
encoding/json.(*encodeState).reflectValue(0xc20806f080, 0x8903c0, 0xc2080d16e0, 0x57)
        /usr/lib/go/src/encoding/json/encode.go:299 +0x72
encoding/json.(*encodeState).marshal(0xc20806f080, 0x8903c0, 0xc2080d16e0, 0x0, 0x0)
        /usr/lib/go/src/encoding/json/encode.go:270 +0xca
encoding/json.Marshal(0x8903c0, 0xc2080d16e0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/lib/go/src/encoding/json/encode.go:135 +0x92
github.com/letsencrypt/boulder/sa.BoulderTypeConverter.ToDb(0x8903c0, 0xc2080d16e0, 0x0, 0x0, 0x0, 0x0)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/sa/type-converter.go:26 +0x110
github.com/letsencrypt/boulder/sa.(*BoulderTypeConverter).ToDb(0xf69740, 0x8903c0, 0xc2080d16e0, 0x0, 0x0, 0x0, 0x0)
        <autogenerated>:1 +0xcf
github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp%2ev1.bindPlan.createBindInstance(0xc20806e160, 0xab, 0xc20806b900, 0xa, 0x10, 0xc20829e0e0, 0x1, 0x1, 0xa0c450, 0x7, ...)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1/gorp.go:307 +0x7ea
github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp%2ev1.(*TableMap).bindUpdate(0xc208034d80, 0x982400, 0xc20806ce10, 0xd9, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1/gorp.go:454 +0xe4d
github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp%2ev1.update(0xc20802c240, 0x7f47085da5e8, 0xc2080d0ec0, 0xc2080f8700, 0x1, 0x1, 0x4a3b4f, 0x0, 0x0)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1/gorp.go:1939 +0x2b5
github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp%2ev1.(*Transaction).Update(0xc2080d0ec0, 0xc2080f8700, 0x1, 0x1, 0x1, 0x0, 0x0)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1/gorp.go:1201 +0x95
github.com/letsencrypt/boulder/sa.(*SQLStorageAuthority).UpdatePendingAuthorization(0xc2080d1020, 0xc20829d230, 0x2b, 0xc20829f078, 0x3, 0xc20829f080, 0x7, 0x1, 0xc20829f048, 0x7, ...)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/sa/storage-authority.go:410 +0x7f9
github.com/letsencrypt/boulder/ra.(*RegistrationAuthorityImpl).UpdateAuthorization(0xc20802c2a0, 0xc20829d230, 0x2b, 0xc20829f078, 0x3, 0xc20829f080, 0x7, 0x1, 0xc20829f048, 0x7, ...)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/ra/registration-authority.go:395 +0x350
github.com/letsencrypt/boulder/wfe.(*WebFrontEndImpl).challenge(0xc20806a400, 0xc20829d230, 0x2b, 0xc20829f078, 0x3, 0xc20829f080, 0x7, 0x1, 0xc20829f048, 0x7, ...)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/wfe/web-front-end.go:775 +0x10b1
github.com/letsencrypt/boulder/wfe.(*WebFrontEndImpl).Authorization(0xc20806a400, 0x7f47085da128, 0xc2080db7c0, 0xc2080b9c70)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/wfe/web-front-end.go:927 +0x5f2
github.com/letsencrypt/boulder/wfe.*WebFrontEndImpl.Authorization·fm(0x7f47085da128, 0xc2080db7c0, 0xc2080b9c70)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/wfe/web-front-end.go:140 +0x45
net/http.HandlerFunc.ServeHTTP(0xc20844efa0, 0x7f47085da128, 0xc2080db7c0, 0xc2080b9c70)
        /usr/lib/go/src/net/http/server.go:1265 +0x41
net/http.(*ServeMux).ServeHTTP(0xc208442f90, 0x7f47085da128, 0xc2080db7c0, 0xc2080b9c70)
        /usr/lib/go/src/net/http/server.go:1541 +0x17d
main.func·001(0x7f47085da128, 0xc2080db7c0, 0xc2080b9c70)
        /home/jakub/dev/letsencrypt/boulder/src/github.com/letsencrypt/boulder/cmd/boulder/main.go:40 +0x111
net/http.HandlerFunc.ServeHTTP(0xc2084526c0, 0x7f47085da128, 0xc2080db7c0, 0xc2080b9c70)
        /usr/lib/go/src/net/http/server.go:1265 +0x41
net/http.serverHandler.ServeHTTP(0xc20843b920, 0x7f47085da128, 0xc2080db7c0, 0xc2080b9c70)
        /usr/lib/go/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc2080db720)
        /usr/lib/go/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
        /usr/lib/go/src/net/http/server.go:1751 +0x35e
```

jsha commented:
Please submit your go-jose changes upstream to be reviewed in parallel with this pull request.

jsha commented on core/objects.go:
```diff
@@ -434,6 +384,30 @@ func (jb *JSONBuffer) UnmarshalJSON(data []byte) (err error) {
+	return jose.JsonWebSignature(jws).Verify(jwk)
+}
+
+// MarshalJSON encodes a JSONBuffer for transmission.
```
Comment is incorrect.

jsha commented on core/objects.go:
```diff
@@ -434,6 +384,30 @@ func (jb *JSONBuffer) UnmarshalJSON(data []byte) (err error) {
+}
+
+// MarshalJSON encodes a JSONBuffer for transmission.
+func (jws AcmeJWS) MarshalJSON() (result []byte, err error) {
```
These functions should be part of go-jose rather than boulder.

jsha commented on va/validation-authority.go:
```diff
@@ -52,21 +55,48 @@ type verificationRequestEvent struct {
+	}
+
+	if len(parsedResponse) != len(target) {
+		return fmt.Errorf("Validation payload did not have all fields")
```
This error message is incorrect: The payload might have had too many fields also.

jmhodges commented on Godeps/_workspace/src/github.com/square/go-jose/jws.go:
```diff
@@ -47,9 +47,9 @@ type JsonWebSignature struct {
 // Signature represents a single signature over the JWS payload and protected header.
 type Signature struct {
 	Header    JoseHeader
+	Signature []byte
```
This doesn't appear to be in the go-jose codebase. If we're forking go-jose, let's move it out of Godeps so that the next person that needs to update go-jose doesn't stomp on the changes.

Upstreaming this, of course, is preferred.

jmhodges commented on core/challenges.go:
```diff
@@ -24,20 +18,10 @@ func SimpleHTTPChallenge() Challenge {
 		Status: StatusPending,
-		R:      RandomString(32),
-		Nonce:  hex.EncodeToString(nonce),
+		Token:  NewToken(),
```
These changes would be easier to review if you listed which spec changes were involved. Neither the pull request nor the original issue mention them. We have commits for them, so maybe they would be instructive (even a range of commits).

jmhodges commented on core/objects.go:
```diff
@@ -239,17 +234,14 @@ type Challenge struct {
 	URI AcmeURL `json:"uri"`
 
-	// Used by simpleHTTP, recoveryToken, and dns challenges
+	// Used by simpleHttp, dvsni, and dns challenges
```
These are comments are great. Thank you.

jmhodges commented on core/objects.go:
```diff
@@ -352,16 +302,16 @@ func (ch Challenge) IsSane(completed bool) bool {
+	// If "tls" is not provided, default to "true"
 	if resp.TLS != nil {
 		ch.TLS = resp.TLS
+	} else {
```
These extra field checks are getting pretty hairy.

Can we adjust the spec to require sub-objects for each different challenge type?

bifurcation replied:
Not sure what you mean here; could you elaborate?  In any case, it doesn't sound like an issue for this PR.

jmhodges commented on core/objects_test.go:
```diff
@@ -40,71 +40,32 @@ func TestRegistrationUpdate(t *testing.T) {
-	test.Assert(t, !chall.IsSane(false), "IsSane should be false")
-	chall.TLS = &tls
-	test.Assert(t, chall.IsSane(false), "IsSane should be true")
+	types := []string{ChallengeTypeSimpleHTTP, ChallengeTypeDVSNI, ChallengeTypeDNS}
```
yay thank you

bifurcation commented:
OK, I've updated this to address the comments above:

* Moved `go-jose` changes to `github.com/letsencrypt/go-jose` and ran `godep update`
* Cleaned up `Challenge.MergeResponse()` to be more sane
* Updated validation tests added while this was pending so that they deal properly with the new challenge format

Ready for re-review.

kuba commented:
Still failing with `Tried to serialize a nil pointer.`

jsha commented:
@kuba: I haven't been able to reproduce so far. What branch are you testing from on the letsencrypt client side?

kuba commented:
https://github.com/kuba/letsencrypt/tree/update-challenges

bifurcation commented:
I was able to reproduce Kuba's branch.  It came down to (1) the le/le client not sending a protected header on its JWS, and (2) `go-jose` being OK with verifying such a JWS, but not OK with serializing it.  I patched `go-jose` to just omit the protected header when there's not one there, and pulled that change into boulder.

jmhodges commented:
LGTM

bifurcation commented:
Thanks for the re-review, @jmhodges.  Merging based on this and earlier reviews.

jmhodges commented:
This is not okay to merge. Earlier reviews are not relevant. There have been substantial change since them.

jcjones commented on core/objects.go:
```diff
@@ -358,17 +313,24 @@ func (ch Challenge) IsSane(completed bool) bool {
+			ch.TLS = resp.TLS
+		} else {
+			ch.TLS = new(bool)
+			*ch.TLS = true
```
Note: I'd love to get a brief explanation of why this is necessary, but I'm sure that's just a gap in my Go knowledge.

bifurcation replied:
Just a simplification.  Treat the different types of challenges independently instead of requiring the reader to tease out what applies to what.

jcjones commented on va/validation-authority.go:
```diff
@@ -59,6 +62,43 @@ type verificationRequestEvent struct {
 	Error        string         `json:",omitempty"`
 }
 
+// TODO Update jws.go to accept jose.JsonWebKey in newVerifier
```
Per [wiki/Boulder-Development](https://github.com/letsencrypt/boulder/wiki/Boulder-Development) this should be in the form of either `TODO(rlb...)` or `TODO(Issue #...)`

bifurcation replied:
I'm just going to delete it, because it doesn't even make sense to me now that I look at it.

jcjones commented on va/validation-authority_test.go:
```diff
@@ -62,25 +68,39 @@ const pathMoved = "301"
 const pathUnsafe = "%"
 const pathUnsafe302 = "302-to-unsafe"
```
This is now unused; is that intentional?

jcjones commented on va/validation-authority_test.go:
```diff
@@ -62,25 +68,39 @@ const pathMoved = "301"
+				currentToken = pathFound
+			}
 			http.Redirect(w, r, pathMoved, 302)
-		} else if strings.HasSuffix(r.URL.Path, pathUnsafe302) {
```
Deliberate removal? If so, the `pathUnsafe` consts should go.

bifurcation replied:
Yes.  The path used is now the token, which is issued by the server.  So it can't have unsafe characters.  I'll delete the `pathUnsafe` consts.

jcjones commented:
All of my comments are either tests, or a comment-comment, so r=jcjones with the comments addressed. We can do a fast-followup PR to apply those changes.

jmhodges commented:
LGTM

jmhodges commented:
Those were inline with JC's comments and very small. This PR is already closed so hey okay let's merge

rolandshoemaker commented:
Post merge r+
